### ISSUE TITLE
Passing localstorage is necessary while using supabase-js in a deno runtime 

### ISSUE THREAD

**Issue by nohehf** [supabase/supabase-js#684]
# Improve documentation

## Link 

[This page could be improved](https://supabase.com/docs/reference/javascript/initializing)

## Describe the problem

While using supabase-js using deno, I ran into an issue: the client was not saving the session. I figured out this comes from the client meant to be used in a browser and then using local storage. Indeed manually passing the storage while initializing the client fixed the issue:
```TS
const supabase = createClient(env.API_URL, env.SUPABASE_KEY, {
  auth: {
    storage: localStorage,
  },
});
```
I struggled to find the solution and had to test things based on function signatures, and other people would like to be able to use the client on a server.

## Describe the improvement

The above page should mention this and provide a solution for Deno and nodejs. Tho for nodejs a third-party package might be necessary.

## Related issues
I believe supabase/gotrue-js#771 is related to the same issue, tho with nodejs.

**Comment by j4w8n** [comment]
You should also be able to set the client's `auth` > `persistSession` option as `false`. This will cause the session to be stored in memory.

Either way, you are correct - it needs some documentation.

**Comment by nohehf** [comment]
Yes, indeed, and this is a particularly useful (and probably widely needed) use-case for edge functions performing actions on the DB.

**Comment by antick** [comment]
> You should also be able to set the client's `auth` > `persistSession` option as `false`. This will cause the session to be stored in memory.
> 
> Either way, you are correct - it needs some documentation.

Setting persistSession to false doesn't work. It is still not creating the in-memory session.

**Comment by j4w8n** [comment]
> > You should also be able to set the client's `auth` > `persistSession` option as `false`. This will cause the session to be stored in memory.
> > 
> > Either way, you are correct - it needs some documentation.
> 
> Setting persistSession to false doesn't work. It is still not creating the in-memory session.

What makes you say this? Unless code has recently changed, it's one or the other. If `true`, it attempts to use local storage; if `false`, it uses memory via variables in the client.

**Comment by antick** [comment]
Hey, I have been struggling with this problem for hours and setting that flag to false alone didn't make any difference.

This worked for me though:

```js
this.clientInstance = createClient(
  this.configService.get('SUPABASE_URL'),
  this.configService.get('SUPABASE_KEY'),
  {
    global: {
      headers: {
        Authorization: this.request.headers.authorization,
      }
    }
  }
);
```

**Comment by j4w8n** [comment]
Glad you got it working.

There are some links to examples, which talk about passing global headers.

https://supabase.com/docs/guides/functions
👇
https://github.com/supabase/supabase/blob/master/examples/edge-functions/supabase/functions/select-from-table-with-auth-rls/index.ts#L26

**Comment by Manouchehri** [comment]
For Cloudflare Workers, this was the solution:

```typescript
const supabase_client = createClient(c.env.SUPABASE_URL, c.env.SUPABASE_KEY, {
  auth: {
    persistSession: false
  }
})
```

Thanks to https://github.com/supabase/supabase-js/issues/684#issuecomment-1376620014 for explaining it! (Just writing this comment so that I and others will find it indexed on Google in the future.)

**Comment by Dramelac** [comment]
Hello

I'm not sure if the problem is related to this issue as none of the solutions above solved the problem (`persistSession: false`, `storage: false` etc) but when I try to retrieve the `await supabaseClient.auth.mfa.getAuthenticatorAssuranceLevel()` I get this;
```js
{
  currentLevel: null,
  nextLevel: null,
  currentAuthenticationMethods: []
}
```
(Even though the MFA is properly configured for the current user)

Here is my code from a supabase edge function. Run locally `Using supabase-edge-runtime-1.67.2 (compatible with Deno v1.45.2)`:
```js
const supabaseClient = createClient<Database, app_schema>(
        Deno.env.get("SUPABASE_URL") ?? "",
        Deno.env.get("SUPABASE_ANON_KEY") ?? "",
        {
            global: {headers: {Authorization: authHeader}},
            // I tried different auth options but nothing fix the getAuthenticatorAssuranceLevel response
            auth: {
                autoRefreshToken: false,
                persistSession: false,
                detectSessionInUrl: false,
                storage: false
            }
        }
    );
// getUser works
const current_user = await supabaseClient.auth.getUser();

// But not MFA policy check
const { data , error } = await supabaseClient.auth.mfa.getAuthenticatorAssuranceLevel()
console.log("Current MFA:", data)

// RLS query works too
[..]
```
If anyone has any idea what the problem is, I'd appreciate it!

**Comment by j4w8n** [comment]
> Hello
> 
> I'm not sure if the problem is related to this issue as none of the solutions above solved the problem (`persistSession: false`, `storage: false` etc) but when I try to retrieve the `await supabaseClient.auth.mfa.getAuthenticatorAssuranceLevel()` I get this;
> 
> {
>   currentLevel: null,
>   nextLevel: null,
>   currentAuthenticationMethods: []
> }
> 
> (Even though the MFA is properly configured for the current user)
> 
> Here is my code from a supabase edge function. Run locally `Using supabase-edge-runtime-1.67.2 (compatible with Deno v1.45.2)`:
> 
> const supabaseClient = createClient<Database, app_schema>(
>         Deno.env.get("SUPABASE_URL") ?? "",
>         Deno.env.get("SUPABASE_ANON_KEY") ?? "",
>         {
>             global: {headers: {Authorization: authHeader}},
>             // I tried different auth options but nothing fix the getAuthenticatorAssuranceLevel response
>             auth: {
>                 autoRefreshToken: false,
>                 persistSession: false,
>                 detectSessionInUrl: false,
>                 storage: false
>             }
>         }
>     );
> // getUser works
> const current_user = await supabaseClient.auth.getUser();
> 
> // But not MFA policy check
> const { data , error } = await supabaseClient.auth.mfa.getAuthenticatorAssuranceLevel()
> console.log("Current MFA:", data)
> 
> // RLS query works too
> [..]
> 
> If anyone has any idea what the problem is, I'd appreciate it!

`getAuthenticatorAssuranceLevel` relies on their being a session in storage, but you aren't setting a session. Setting the global authorization header allows `getUser` to work though.

Here are two options you have:
1. Use `getClaims`. Get rid of your `getUser` call (or maybe that was just an experiment), and call `getClaims(authHeader)`. Note this calls `getUser` internally (to verify the JWT) and will give you back your claims data, which includes what you're looking for. However, it will not return the user like `getUser` would do, if you're still needing that.
    ```js
    const { data, error } = await supabaseClient.auth.getClaims(authHeader)

    /* handle scenario where there's an `error` or `data` is `null` */

    const { claims } = data
    const assurance_level = claims.aal
    ```
2. Verify the `authHeader` yourself, with your JWT secret, then decode the JWT (if not done during verify) and extract the `aal` value from the JWT payload. This does not require a network request to your Supabase instance (which is what `getUser` and `getClaims` invokes).

**Comment by Dramelac** [comment]
> 1. Use `getClaims`. Get rid of your `getUser` call (or maybe that was just an experiment), and call `getClaims(authHeader)`. Note this calls `getUser` internally (to verify the JWT) and will give you back your claims data, which includes what you're looking for. However, it will not return the user like `getUser` would do, if you're still needing that.

Hello,

Thank you @j4w8n for your reply.

This function `await supabaseClient.auth.getClaims(authHeader)` would be very nice but i don't find it in the documentation and my IDE say it doesn't exist ? Am i missing something ?

![Image](https://github.com/user-attachments/assets/ca916cec-2e33-48b9-80c2-4298fbc65c2e)

I also need to know if the current user has enrolled an MFA device, which is why `getAuthenticatorAssuranceLevel()` is my preferred action as it also checks what the next available level is for the current user. 
The `getUser` function list availables factors so it can be used to check this too but as you said, it would mean 2 network request.

**Comment by j4w8n** [comment]
@Dramelac then you can always follow option 3. You can add your JWT secret to the edge functions env vars at https://supabase.com/dashboard/project/_/functions/secrets

Or figure out how to use a newer version of supabase-js in the edge function, so that getClaims is available.

**Comment by Dramelac** [comment]
If I can avoid re-implementing the JWT check myself, I'd prefer that.
Wouldn't it be a good idea to upgrade the `getAuthenticatorAssuranceLevel()` function to provide it with the auth header as a direct parameter, as `getUser(authHeader)` already does, for example?

![Image](https://github.com/user-attachments/assets/ac01114b-97ba-4348-a957-f867a640f877)

**Comment by j4w8n** [comment]
> If I can avoid re-implementing the JWT check myself, I'd prefer that.
> Wouldn't it be a good idea to upgrade the `getAuthenticatorAssuranceLevel()` function to provide it with the auth header as a direct parameter, as `getUser(authHeader)` already does, for example?
> 
> ![Image](https://github.com/user-attachments/assets/ac01114b-97ba-4348-a957-f867a640f877)

That'd be a call the auth team would need to make. You can create an issue for it, to see what they say.
